### PR TITLE
Refactor classes/PrestaShopAutoload.php

### PR DESCRIPTION
**Pull request by mvorisek**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Pure refactoring from another autoloading PR to not polute the changes with non-functional changes. This PR makes `classes/PrestaShopAutoload.php` independed of config constants that do not make sense in this file as if they are defined with any other than non-default changes, the PrestaShop will not work. This PR is needed to further integrate the legacy autoloader into composer autoloader, once merged, I will then finish https://github.com/PrestaShop/PrestaShop/pull/24864 for the `develop` branch.
| Type?             | refacto
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | n/a
| How to test?      | tests must pass
| Possible impacts? | no

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/28121)

### CODE REVIEW

File: classes/PrestaShopAutoload.php [hunk 1]
```diff
@@ -57,11 +57,11 @@ class PrestaShopAutoload
 
     protected function __construct()
     {
-        $this->root_dir = _PS_CORE_DIR_ . '/';
+        $this->root_dir = dirname(__DIR__);
```
Comment (jolelievre) [inline classes/PrestaShopAutoload.php:60]:
```suggestion
        $this->root_dir = dirname(__DIR__) . '/';
```

This will avoid adding the slash everywhere else in this class

Comment (jolelievre) [reply classes/PrestaShopAutoload.php:60]: However I don't understand `dirname(__DIR__)` is supposed to be equal to the `classes` folder Not the root folder so it seems incorrect unless I'm missing something

File: classes/PrestaShopAutoload.php [hunk 2]
```diff
@@ -81,14 +81,24 @@ public static function getInstance()
...
     }
 
+    /**
+     * Get cache directory.
+     *
+     * @return string
+     */
+    protected static function getCacheDir()
+    {
+        return dirname(__DIR__) . '/cache';
```
Comment (jolelievre) [inline classes/PrestaShopAutoload.php]: Same here cache dir is supposed to be `var/cache/dev` (for dev environment) Here this seems to target `classes/cache`

File: classes/PrestaShopAutoload.php [hunk 3]
```diff
@@ -57,11 +57,11 @@ class PrestaShopAutoload
 
     protected function __construct()
     {
-        $this->root_dir = _PS_CORE_DIR_ . '/';
+        $this->root_dir = dirname(__DIR__);
         $file = static::getCacheFileIndex();
         $stubFile = static::getStubFileIndex();
         if (@filemtime($file) && is_readable($file) && @filemtime($stubFile) && is_readable($stubFile)) {
-            $this->index = include $file;
+            $this->index = require $file;
```
Comment (jolelievre) [inline classes/PrestaShopAutoload.php:64]: Why is this change needed?

Comment (eternoendless) [reply classes/PrestaShopAutoload.php:64]: It makes sense to me to make it fail instead of issuing a warning. If the cache file exists, it better be valid.

File: classes/PrestaShopAutoload.php (hunk 2 above)
Comment (eternoendless) [reply classes/PrestaShopAutoload.php]: +1

File: classes/PrestaShopAutoload.php (hunk 1 above)
Comment (mvorisek) [reply classes/PrestaShopAutoload.php:60]: > This will avoid adding the slash everywhere else in this class

that is correct, but the convenction is normally to not add the `/` to a dir variable, which is also safer, as user can set the path /wo the trailing slash.

Comment (mvorisek) [reply classes/PrestaShopAutoload.php:60]: > However I don't understand `dirname(__DIR__)` is supposed to be equal to the `classes` folder Not the root folder so it seems incorrect unless I'm missing something

image

`_PS_CORE_DIR_` is the PrestaShop/project root directory

File: classes/PrestaShopAutoload.php (hunk 2 above)
Comment (mvorisek) [reply classes/PrestaShopAutoload.php]: image

yes, it is correct that for dev environment the `_PS_CACHE_DIR_` was different, but with https://github.com/PrestaShop/PrestaShop/pull/24864/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R174 it is impossible to detect the dev environment easily, as this file is loaded before any class is usable, so `_PS_CACHE_DIR_` constant set in:

image

cannot be used

is it ok to have one cache location for prod & dev for autoloader?

File: classes/PrestaShopAutoload.php [hunk 4]
```diff
@@ -81,14 +81,24 @@ public static function getInstance()
         return static::$instance;
     }
 
+    /**
+     * Get cache directory.
+     *
+     * @return string
+     */
+    protected static function getCacheDir()
```
Comment (Progi1984) [inline classes/PrestaShopAutoload.php]:
```suggestion
    protected static function getCacheDir(): string
```

**Review by Progi1984 — APPROVED** [review]

**Comment by mvorisek** [comment]
Guys, this PR is needed for https://github.com/PrestaShop/PrestaShop/pull/28129, can it be merged soon? Thanks.

File: classes/PrestaShopAutoload.php (hunk 1 above)
Comment (MeKeyCool) [reply classes/PrestaShopAutoload.php:60]: @jolelievre , I agree with your comment, it is disturbing cause `__DIR__` is current php file directory and `dirname` gives the parent directory of this directory xD
So this `dirname(__DIR__)` is like a `/..` assuming that current autoload file is at two levels from `_PS_CORE_DIR_` which is project root directory ...

:thinking: 
I don't like this change cause it is like changing configuration to hardcode.
I can accept that autoloading is based on an know directory tree for classes and I don't know the full Prestashop to understand if this change is good or not.

Is it the only autoload point ?
Has it was configuration and now it is hardcoded, what happens for other files that use `_PS_CORE_DIR_` ?
Do you assume that this variable isn't the same meaning than this usage for autoload file ?
(imagine we change the project structure for example, should we update several hardcoded files ?)

File: classes/PrestaShopAutoload.php [hunk 5]
```diff
@@ -81,14 +81,24 @@ public static function getInstance()
...
     }
 
+    /**
+     * Get cache directory.
+     *
+     * @return string
+     */
+    protected static function getCacheDir(): string
+    {
+        return dirname(__DIR__) . '/var/cache';
```
Comment (MeKeyCool) [inline classes/PrestaShopAutoload.php:91]: Ok, I can understand that you don't want to use config/environment variable for autoloading stuff (finding files under project defined directory tree).

But I really think hard coding cache directory isn't a good idea.
Cache is a dynamic directory that you may want to control when you are devops/admin (for example if you want to create some persistent volumes in a docker container). And the thing you don't want is to mix your sources with dynamic data.
For this question, Prestashop is already a mess. Removing `_PS_CACHE_DIR_` variable turns any refactoring even harder ...

File: classes/PrestaShopAutoload.php [hunk 6]
```diff
@@ -222,7 +229,7 @@ public function generateIndex()
 
         // Write classes index on disc to cache it
         $filename = static::getCacheFileIndex();
-        @mkdir(_PS_CACHE_DIR_, 0777, true);
+        @mkdir(static::getCacheDir(), 0777, true);
```
Comment (MeKeyCool) [inline classes/PrestaShopAutoload.php:232]: Hmmm ... I am disturbed by this line ... Does this mean that we create a cache file that should help application to load classes and we put `777` rights on it ?

Doesn't it imply that anybody that is able to access this file may edit it and then completely take control of Prestashop ? (I mean changing autoloading behavior looks even stronger than a full db access)

Even if it is not a part of your Pull Request, I would like to ask confirmations to @PrestaShop/prestashop-core-developers because if it is what I understood, this line is bad enough to justify a security fix on all Prestashop we can ^^'

File: tests/Integration/Classes/PrestaShopAutoloadTest.php [hunk 7]
```diff
@@ -82,12 +82,12 @@ class Connection extends ConnectionCore {}'
         PrestaShopAutoload::getInstance()->generateIndex();
         $this->assertFileExists($this->file_index);
         $data = include $this->file_index;
-        $this->assertEquals($data['OrderControllerCore']['path'], 'controllers/front/OrderController.php');
-        $this->assertEquals($data['Connection']['override'], false);
+        $this->assertSame('controllers/front/OrderController.php', $data['OrderControllerCore']['path']);
+        $this->assertFalse($data['Connection']['override']);
```
Comment (MeKeyCool) [inline tests/Integration/Classes/PrestaShopAutoloadTest.php:86]: :+1:

**Review by MeKeyCool — CHANGES_REQUESTED** [review]
I am not sure to well understand what is you strategy to remove legacy autoloader and to use composer one only but I am really interested by the subject ^^
If you have more information to share, I would enjoy to know more :)

File: classes/PrestaShopAutoload.php (hunk 5 above)
Comment (mvorisek) [reply classes/PrestaShopAutoload.php:91]: hardcoded path is needed, as this file is included before anything else, even before any configuration is/can be loaded

Comment (MeKeyCool) [reply classes/PrestaShopAutoload.php:91]: If I understand well what you say, functions like this one 
```php
public static function getCacheFileIndex()
{
    return _PS_CACHE_DIR_ . 'class_index.php';
}
```
would search `class_index.php` in a wrong location during some process ?
Could you describe the use case where you have seen this behavior/bug ?

Comment (mvorisek) [reply classes/PrestaShopAutoload.php:91]: please see dependent PR https://github.com/PrestaShop/PrestaShop/pull/28129/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R182 - `classes/PrestaShopAutoload.php` is to be loaded before any config read, thus no user defined constants/config can be used

File: classes/PrestaShopAutoload.php (hunk 6 above)
Comment (MeKeyCool) [reply classes/PrestaShopAutoload.php:232]: :thinking: Maybe my thought are stupid :thinking: This line concerns only a directory ... :thinking:

Comment (mvorisek) [reply classes/PrestaShopAutoload.php:232]: I belive 0777 is correct here, as PrestaShop may be installed via CLI (or run using php integrated webserver for example), eg. easily as a different user, anything else than 0777 will render the PrestaShop unusable.

File: classes/PrestaShopAutoload.php (hunk 5 above)
Comment (MeKeyCool) [reply classes/PrestaShopAutoload.php:91]: Ok I got it.
So in fact you have anticipated a bootstrap process / architecture and your argue is that you have to implement this way cause of your goal.
Is it right to say that you want to remove environment variables so you have to hard code everything ?

Then I think we don't agree your goal/vision.

Why do you want to remove environment variable ?
Do you think Prestashop directory structure is right ? Once you have hard coded everything, what is your strategy to make it evolve ?
You didn't answer my implicit question about modern usages like Docker context with custom persistent volumes and read/write management issues ...

:thinking: 
Furthermore, I am thinking that we may not see Prestashop the same way neither :thinking:
What is Prestashop for you ? Do you see it as a standard website ?

Comment (MeKeyCool) [reply classes/PrestaShopAutoload.php:91]: I am curious about the opinion of more experienced developers @PrestaShop/prestashop-core-developers  :slightly_smiling_face:

Comment (mvorisek) [reply classes/PrestaShopAutoload.php:91]: > Is it right to say that you want to remove environment variables so you have to hard code everything ?

strictly hardcode only the path of autoloader index cache (AFAK only one file on generated on the 1st run (or if deleted after module install/uninstall))

**Comment by mvorisek** [comment]
This PR is needed for https://github.com/PrestaShop/PrestaShop/pull/28129, can someone please review it

**Comment by mvorisek** [comment]
This PR is needed for https://github.com/PrestaShop/PrestaShop/pull/28129, can someone please review it

**Comment by mvorisek** [comment]
It seems this PR has been forgotten, can it be reviewed and merged soon? Thanks in advance!

**Comment by mvorisek** [comment]
Can some maintainer please get onto this PR?

**Comment by Progi1984** [comment]
Ping @PrestaShop/prestashop-maintainers & @PrestaShop/committers

**Comment by FabienPapet** [comment]
hey @mvorisek ! I just created a PR #29923  which moves out the autoload as a separated package, you can find more details here

**Comment by mvorisek** [comment]
@FabienPapet thank you for pinging me - please see https://github.com/PrestaShop/PrestaShop/pull/24864, you should integrate the autoload to composer, no explicit registering should be needed (if needed, add it to always included file via composer.json)

**Comment by FabienPapet** [comment]
Hello @mvorisek 

Thank you for your work of this PR, the whole autoload feature has been rewritten in https://github.com/PrestaShop/PrestaShop/pull/29923 and contains these fixes.

I'm now closing this as the new autoload contains these fixes.
